File: core/src/main/java/uk/co/amlcurran/social/EventsService.java
package uk.co.amlcurran.social;

import java.util.List;

import javax.annotation.Nonnull;

import uk.co.amlcurran.social.core.SparseArray;

public class EventsService {
    private final TimeRepository timeRepository;
    private final EventsRepository eventsRepository;
    private final TimeCalculator timeCalculator;

    public EventsService(TimeRepository dateCreator, EventsRepository eventsRepository, TimeCalculator timeCalculator) {
        this.eventsRepository = eventsRepository;
        this.timeRepository = dateCreator;
        this.timeCalculator = timeCalculator;
    }

    @Nonnull
    public CalendarSource getCalendarSource(final int numberOfDays, final Timestamp now) {
        Timestamp nowTime = timeCalculator.startOfToday();
        Timestamp nextWeek = nowTime.plusDays(numberOfDays);
        TimeOfDay fivePm = timeRepository.borderTimeStart();
        TimeOfDay elevenPm = timeRepository.borderTimeEnd();

        List<CalendarItem> calendarItems = eventsRepository.getCalendarItems(nowTime, nextWeek, fivePm, elevenPm);

        SparseArray<CalendarSlot> itemArray = new SparseArray<>(numberOfDays);
        int epochToNow = now.daysSinceEpoch();
        for (CalendarItem item : calendarItems) {
            int key = item.startTime().daysSinceEpoch() - epochToNow;
            CalendarSlot slot = itemArray.get(key, new CalendarSlot());
            slot.addItem(item);
            itemArray.put(key, slot);
        }

        return new CalendarSource(itemArray, numberOfDays, timeCalculator);
    }

}

File: android/core/src/main/java/uk/co/amlcurran/social/EventsService.kt
package uk.co.amlcurran.social

data class Event(val item: EventCalendarItem, val location: String)

class EventsService(
        private val timeRepository: TimeRepository,
        private val eventsRepository: EventsRepository,
        private val timeCalculator: TimeCalculator
) {

    fun getCalendarSource(numberOfDays: Int, now: Timestamp): CalendarSource {
        val nowTime = timeCalculator.startOfToday()
        val nextWeek = nowTime.plusDays(numberOfDays)
        val fivePm = timeRepository.borderTimeStart()
        val elevenPm = timeRepository.borderTimeEnd()

        val calendarItems = eventsRepository.getCalendarItems(nowTime, nextWeek, fivePm, elevenPm)

        val itemArray = mutableMapOf<Int, CalendarSlot>()
        val epochToNow = now.daysSinceEpoch()
        for (item in calendarItems) {
            val key = item.startTime.daysSinceEpoch() - epochToNow
            val slot = itemArray[key] ?: CalendarSlot()
            slot.addItem(item)
            itemArray[key] = slot
        }

        return CalendarSource(itemArray, numberOfDays, timeCalculator, timeRepository)
    }

    fun eventWithId(eventId: String): Event? = eventsRepository.event(eventId)

}
